xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -0,0 +1,455 @@
+      ConnectivityState state = getSubchannelStateInfoRef(subchannel).value.getState();
+      if (state == READY) {
+        return state;
+      }
+      if (state == TRANSIENT_FAILURE) {
+        failureCount++;
+      } else if (state == CONNECTING) {
+        hasConnecting = true;
+      } else if (state == IDLE) {
+        hasIdle = true;
+      }
+    }
+    if (failureCount >= 2) {
+      return TRANSIENT_FAILURE;
+    }
+    if (hasConnecting) {
+      return CONNECTING;
+    }
+    return hasIdle ? IDLE : TRANSIENT_FAILURE;
+  }
+
+  private static void shutdownSubchannel(Subchannel subchannel) {
+    subchannel.shutdown();
+    getSubchannelStateInfoRef(subchannel).value = ConnectivityStateInfo.forNonError(SHUTDOWN);
+  }
+
+  /**
+   * Converts list of {@link EquivalentAddressGroup} to {@link EquivalentAddressGroup} set and
+   * remove all attributes. The values are the original EAGs.
+   */
+  private static Map<EquivalentAddressGroup, EquivalentAddressGroup> stripAttrs(
+      List<EquivalentAddressGroup> groupList) {
+    Map<EquivalentAddressGroup, EquivalentAddressGroup> addrs =
+        new HashMap<>(groupList.size() * 2);
+    for (EquivalentAddressGroup group : groupList) {
+      addrs.put(stripAttrs(group), group);
+    }
+    return addrs;
+  }
+
+  private static EquivalentAddressGroup stripAttrs(EquivalentAddressGroup eag) {
+    return new EquivalentAddressGroup(eag.getAddresses());
+  }
+
+  private static Ref<ConnectivityStateInfo> getSubchannelStateInfoRef(
+      Subchannel subchannel) {
+    return checkNotNull(subchannel.getAttributes().get(STATE_INFO), "STATE_INFO");
+  }
+
+  private static final class RingHashPicker extends SubchannelPicker {
+    private final SynchronizationContext syncContext;
+    private final List<RingEntry> ring;
+    // Avoid synchronization between pickSubchannel and subchannel's connectivity state change,
+    // freeze picker's view of subchannel's connectivity state.
+    private final Map<EquivalentAddressGroup, SubchannelView> pickableSubchannels;  // read-only
+
+    private RingHashPicker(
+        SynchronizationContext syncContext, List<RingEntry> ring,
+        Map<EquivalentAddressGroup, Subchannel> subchannels) {
+      this.syncContext = syncContext;
+      this.ring = ring;
+      pickableSubchannels = new HashMap<>(subchannels.size());
+      for (Map.Entry<EquivalentAddressGroup, Subchannel> entry : subchannels.entrySet()) {
+        Subchannel subchannel = entry.getValue();
+        ConnectivityStateInfo stateInfo = subchannel.getAttributes().get(STATE_INFO).value;
+        pickableSubchannels.put(entry.getKey(), new SubchannelView(subchannel, stateInfo));
+      }
+    }
+
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      Long requestHash = args.getCallOptions().getOption(XdsNameResolver.RPC_HASH_KEY);
+      if (requestHash == null) {
+        return PickResult.withError(RPC_HASH_NOT_FOUND);
+      }
+      if (ring.isEmpty()) {
```

Comment:
How would this be possible? It isn't clear to me why this would choose to queue.

Reply:
Right, this will never happen. A `RingHashPicker` won't be created if not having any backend addresses. Deleted this line.